### PR TITLE
Drag & drop

### LINKED ISSUE #61: Implement drag & drop

### PR DESCRIPTION
**Pull request by dnarcese**
Added drag & drop as well as a delete file confirmation.

Closes #61

### CODE REVIEW

File: frontend/src/upload.js [hunk 1]
```diff
@@ -62,25 +89,33 @@ $(document).ready(function() {
...
       link.innerHTML = url;
       localStorage.setItem(info.fileId, info.deleteToken);
-      let del = document.createElement('td');
-      let btn = document.createElement('button');
-      btn.innerHTML = 'x';
-      btn.classList.add('delete-btn');
-      btn.addEventListener('click', e => {
+
+      // delete file
+      console.log($popupText.first());
```
Comment (ericawright) [inline frontend/src/upload.js]: nit: remove the console.logs

File: frontend/src/upload.js [hunk 2]
```diff
@@ -62,25 +89,33 @@ $(document).ready(function() {
...
+      $popupText.find('.del-file').click(e => {
         FileSender.delete(
           info.fileId,
           localStorage.getItem(info.fileId)
         ).then(() => {
-          e.target.parentNode.parentNode.remove();
+          e.target.parentNode.parentNode.parentNode.parentNode.remove();
```
Comment (ericawright) [inline frontend/src/upload.js]: With jQuery you could try `$(e.target).parents('tr')`
with vanilla JS `e.target.closest('tr')`

File: frontend/src/upload.js [hunk 3]
```diff
@@ -62,25 +89,33 @@ $(document).ready(function() {
...
-      row.appendChild(del);
+
+      // show popup
+      btn.addEventListener('click', e => {
+        $popupText.toggleClass('show');
+      });
+      // hide popup
+      $popupText.find('.nvm').click(e => {
+        $popupText.toggleClass('show');
+      });
```
Comment (ericawright) [inline frontend/src/upload.js]: these two have the same fn after the click event, perhaps we name the fn and reuse it?

File: views/index.handlebars [hunk 4]
```diff
@@ -41,7 +41,9 @@
           <th width=18%>Expires in</th>
           <th width=7%>Delete</th>
         </tr>
-        <div data-role="popup" id="popupArrow" data-arrow="true">
+        <!-- <div class="popup">
```
Comment (ericawright) [inline views/index.handlebars]: nit: remove commented code

File: frontend/src/upload.js [hunk 5]
```diff
@@ -29,25 +33,48 @@ $(document).ready(function() {
...
+  // on file upload by browse or drag & drop
+  window.onUpload = event => {
+    event.preventDefault();
+    let file = '';
+    if (event.type == 'drop') {
+      file = event.dataTransfer.files[0];
+    } else {
+      file = event.target.files[0];
+    }
     let fileList = $('#uploaded-files');
```
Comment (ericawright) [inline frontend/src/upload.js]: nit `$` in front to follow jquery convention

File: public/main.css [hunk 6]
```diff
@@ -108,6 +108,59 @@ td {
...
+   background-color: #555;
+   color: #fff;
+   text-align: center;
+   border-radius: 6px;
+   padding: 8px 0;
+   position: absolute;
+   z-index: 1;
+   bottom: 125%;
+   left: 50%;
+   margin-left: -80px;
```
Comment (ericawright) [inline public/main.css]: do you need both left and margin-left here?

File: public/main.css [hunk 7]
```diff
@@ -108,6 +108,59 @@ td {
...
+   margin-left: -5px;
+   border-width: 5px;
+   border-style: solid;
+   border-color: #555 transparent transparent transparent;
+}
+
+/* Toggle this class when clicking on the popup container (hide and show the popup) */
+.popup .show {
+   visibility: visible;
+   -webkit-animation: fadeIn 1s;
```
Comment (ericawright) [inline public/main.css]: shouldn't need `-webkit`, it's old syntax
Can we change this to a transition, my understanding is that it is more efficient than a keyframes animation.

File: frontend/src/upload.js [hunk 8]
```diff
@@ -29,25 +33,48 @@ $(document).ready(function() {
...
 
-  let onChange = event => {
-    const file = event.target.files[0];
-
+  // on file upload by browse or drag & drop
+  window.onUpload = event => {
+    event.preventDefault();
+    let file = '';
+    if (event.type == 'drop') {
+      file = event.dataTransfer.files[0];
```
Comment (pdehaan) [inline frontend/src/upload.js]: Does this work if I drop multiple files on the drop target from Finder? Or will it only handle the first image?

File: server/portal_server.js [hunk 9]
```diff
@@ -100,6 +100,7 @@ app.post('/delete/:id', (req, res) => {
     .then(err => {
       if (!err) {
         console.log('Deleted.');
+        res.sendStatus(200);
```
Comment (dannycoates) [inline server/portal_server.js]: nice catch :)

File: frontend/src/upload.js (hunk 8 above)
Comment (dannycoates) [inline frontend/src/upload.js]: Good question. If it doesn't we can follow up with a solution in another PR

Comment (pdehaan) [inline frontend/src/upload.js]: Spoiler alert: It doesn't.

But it looks like we can do `Array.from().forEach()` (or just a regular `for` loop) on the FileList as a workaround. But that exposes another dull edge case where you're dragging 2-3 files onto the app and it returns you a single share URL that you can copy to your clipboard.

```js
  window.onUpload = event => {
    event.preventDefault();
    let fileList = $('#uploaded-files');
    let files = [];

    if (event.type === 'drop') {
      files = event.dataTransfer.files;
    } else {
      files = event.target.files;
    }

    Array.from(files).forEach(file => { ... });
```

File: frontend/src/upload.js [hunk 10]
```diff
@@ -63,25 +90,37 @@ $(document).ready(function() {
...
       });
-      del.appendChild(btn);
-      row.appendChild(del);
+
+      // show popup
+      btn.addEventListener('click', e => {
+        toggleShow();
+      });
+      // hide popup
+      $popupText.find('.nvm').click(e => {
```
Comment (ericawright) [inline frontend/src/upload.js]: btn.addEventListener('click', toggleShow);

      $popupText.find('.nvm').click(toggleShow);

**Comment by ericawright** [comment]
just the one more nit and then I think we're good to merge :)

**Pull request merged**